### ISSUE TITLE
Gibberish output after 70k steps

### ISSUE THREAD

**Issue by godspirit00** [VinAIResearch/XPhoneBERT#4]
Hello,
I tried training VITS with XPhoneBERT as in this repo using my own dataset. I processed the dataset (which is already normalized) as instructed in README by using spaCy to segment words, and preprocessed it with `preprocess.py`.
After 70k, it only produces some gibberish output like the following one,

https://github.com/VinAIResearch/XPhoneBERT/assets/16176151/6f179142-9e30-4fbe-be52-9dca1f928f5f

What am I missing?
Thank you!

**Comment by thelinhbkhn2014** [comment]
It depends on the number of training steps you set in the config file. If the number of training steps you set in the config file is 300K, this can be understood because we froze XPhoneBERT in the first 25% of training steps to limit the issue of hyperparameters being sensitive to different languages. You may get better results when XPhoneBERT starts to be finetuned.
Another solution for you is to comment out lines 121 to 128 in the train.py file to finetune XPhoneBERT on the whold training process. However, it may be sensitive to params for different datasets or languages.
If you have any question, please feel free to ask me. Thanks for your interest!

**Comment by godspirit00** [comment]
Thanks for your help! I'll continue with the training.

**Comment by datquocnguyen** [comment]
@godspirit00 is it better now?

**Comment by godspirit00** [comment]
@datquocnguyen Sorry for the late reply.
Unfortunately, no.
I set the total number of epochs to 1000. It is now at Epoch 290 / 500k steps, and the output is still gibberish.

https://github.com/VinAIResearch/XPhoneBERT/assets/16176151/9772e7dd-7409-4e44-82f9-f5c694431b30

The duration is also incorrect. GT is 8s but the gen is 2s.

Not sure if it is related, but the training encountered OOM at around epoch 250, probably when XPhoneBERT began to be finetuned. Then I edited the config and reduced the batch size, then resume the training. The number of steps jumped from 200k to 400k after resuming.
If it's the reason, then I will have to restart the whole training right?

**Comment by datquocnguyen** [comment]
It is likely that you would need to restart the whole training process.

If you decrease the batch size, it is advisable to decrease the initial learning rate accordingly. For example, if you halve the batch size, the learning rate should also be halved (equivalently).

**Comment by godspirit00** [comment]
OK, I will restart the training now. Thank you for your help.

**Comment by cnlinxi** [comment]
> OK, I will restart the training now. Thank you for your help.

Hello, @godspirit00 , how is it now? I am also encountering similar problems. GT is 4s, but generated speech is 1s

**Comment by godspirit00** [comment]
@cnlinxi My training is at epoch 198/1000 now. I guess the finetuning of XPhoneBERT has not started so the output is still gibberish. I am waiting for it to go further.

**Comment by cnlinxi** [comment]
> @cnlinxi My training is at epoch 198/1000 now. I guess the finetuning of XPhoneBERT has not started so the output is still gibberish. I am waiting for it to go further.

@godspirit00 
My training is at epoch 1000/1600, 600 epochs of XPhoneBERT fine-tuning have been made, but the synthesized speech is still gibberish. It seems very difficult to converge, this is so strange.

**Comment by cnlinxi** [comment]
@datquocnguyen @thelinhbkhn2014 
According to the paper, the only difference in model is replacing the encoder of VITS with XPhoneBERT. I follow [README](https://github.com/VinAIResearch/XPhoneBERT/blob/main/VITS_with_XPhoneBERT/README.md) but got bad result: the synthesized speech is gibberish and the predicted duration is not correct. Do you have any idea? Thank you.

![image](https://github.com/VinAIResearch/XPhoneBERT/assets/11946496/f113a4b2-37b6-45ef-9847-98b77967cc8b)


> https://github.com/VinAIResearch/XPhoneBERT/blob/90d278d7d01d02c218eba0537dc60b1f1380c3d1/VITS_with_XPhoneBERT/models.py#L136

**Comment by lexkoro** [comment]
You can try training your model first without xphonebert and once it converges you can replace the character embedding with xphonebert and continue training.

**Comment by datquocnguyen** [comment]
@cnlinxi Which part of the readme are you following: LJSpeech or your own dataset?

**Comment by cnlinxi** [comment]
@datquocnguyen To be precise, it's this part: https://github.com/VinAIResearch/XPhoneBERT/blob/main/VITS_with_XPhoneBERT/README.md#for-another-tts-dataset , I use my own dataset. I use the original VITS for this dataset (phonemes generated by espeak-ng) to ensure it's ok.

**Comment by cnlinxi** [comment]
@lexkoro yes, the training has already 1160/1600, but the generated speech is still gibberish. I think it's time to stop and re-training original VITS with my own dataset (phonemes generated by [CharsiuG2P](https://github.com/lingjzhu/CharsiuG2P/tree/main)) . If the dataset and code are both ok, the original VITS should be convergence for 10k step. I will tell you if the result has come out.

**Comment by godspirit00** [comment]
@cnlinxi My training is at epoch 258/1000 now, and it can now generate reasonable output, though not very good. I'll keep on training for some more epochs to see if it can improve more.

**Comment by datquocnguyen** [comment]
> @datquocnguyen To be precise, it's this part: https://github.com/VinAIResearch/XPhoneBERT/blob/main/VITS_with_XPhoneBERT/README.md#for-another-tts-dataset , I use my own dataset. I use the original VITS for this dataset (phonemes generated by espeak-ng) to ensure it's ok.

Did you use our preprocess.py script to generate phonemes for your own dataset? @cnlinxi

**Comment by datquocnguyen** [comment]
@godspirit00 258/1000 refers to that xphonebert is just fine-tuned for 258 - 25% * 1000 = 8 epochs. I expect you'd get better results when the whole training process is completed.

**Comment by datquocnguyen** [comment]
`python preprocess.py --input_file filelists/training_texts_file.txt --output_file filelists/training_phoneme_sequences.txt --language <ISO-639-3-language-code> --batch_size 64 --cuda`

`python preprocess.py --input_file filelists/validation_texts_file.txt --output_file filelists/validation_phoneme_sequences.txt --language <ISO-639-3-language-code> --batch_size 64 --cuda`

`<ISO-639-3-language-code>` is the corresponding ISO 639-3 code of your own dataset's language. The ISO 639-3 codes of supported languages are available at [HERE](https://github.com/VinAIResearch/XPhoneBERT/blob/main/LanguageISO639-3Codes.md).

**Comment by cnlinxi** [comment]
> > @datquocnguyen To be precise, it's this part: https://github.com/VinAIResearch/XPhoneBERT/blob/main/VITS_with_XPhoneBERT/README.md#for-another-tts-dataset , I use my own dataset. I use the original VITS for this dataset (phonemes generated by espeak-ng) to ensure it's ok.
> 
> Did you use our preprocess.py script to generate phonemes for your own dataset? @cnlinxi

@datquocnguyen 

Yes, I process my own dataset by this script. I have trained original VITS with same dataset and similar config, it can synthesize almost normal result. It's so strange, let me check more.

**trained with original VITS:**

![image](https://github.com/VinAIResearch/XPhoneBERT/assets/11946496/db427990-efb9-47db-9288-1c35ee5885f7)

**trained with XPhoneBERT VITS:**

![image](https://github.com/VinAIResearch/XPhoneBERT/assets/11946496/8883492f-cb53-4435-820a-bc29070b2b46)

**raw text:**

```
after which time a very moderate use of liquor will not be considered a violation. However, all members of the White House Detail
```

**after process:**

```
ˈæ f t ɝ ▁ ˈh w ɪ t ʃ ▁ ˈt a ɪ m ▁ ˈe ɪ ▁ ˈv ɛ ɹ i ▁ ˈm ɑ d ɝ ˌe ɪ t ▁ ˈj u s ▁ ˈə v ▁ ˈɫ ɪ k ɝ ▁ ˈw ɪ ɫ ▁ ˈn ɑ t ▁ b i ▁ k ə n ˈs ɪ d ɝ d ▁ ˈe ɪ ▁ v a ɪ ə ˈɫ e ɪ ʃ ə n ▁ ˌh a ʊ ˈɛ v ɝ ▁ ˈɔ ɫ ▁ ˈm ɛ m b ɝ z ▁ ˈə v ▁ ð ə ▁ ˈh w a ɪ t ▁ ˈh a ʊ s ▁ ˈd i t e ɪ ɫ
```

**Comment by cnlinxi** [comment]
> @cnlinxi My training is at epoch 258/1000 now, and it can now generate reasonable output, though not very good. I'll keep on training for some more epochs to see if it can improve more.

@godspirit00 

I'm glad you've achieved effective results. Maybe there's something I haven't done correctly yet

**Comment by godspirit00** [comment]
@cnlinxi There might be something you're missing. By observing the file lists in this repo, I found that there should also be a space between punctuations and words. This is probably what the authors mean by `word-segmented` in the README.
For example,
`DUMMY/LJ049-0022.wav|the secret service believed that it was very doubtful that any president would ride regularly in a vehicle with a fixed top , even though transparent .`
Instead of 
`DUMMY/LJ049-0022.wav|the secret service believed that it was very doubtful that any president would ride regularly in a vehicle with a fixed to, even though transparent.`
After the preprocessing, the punctuations should still be there. For example, as shown in the README of `text2phonemesequence`, 
the sentence `The overwhelming majority of people in this country know how to sift the wheat from the chaff in what they hear and what they read .` 
will be processed to
`ˈθ i ▁ ˈo ʊ v ɝ ˌw ɛ ɫ m ɪ ŋ ▁ m ə ˈd ʒ ɔ ɹ ə t i ▁ ˈɑ f ▁ ˈp i p ə ɫ ▁ ˈɪ n ▁ ˈθ ɪ s ▁ ˈk a ʊ n t ɹ i ▁ ˈn o ʊ ▁ ˈh o ʊ ▁ ˈt o ʊ ▁ ˈs ɪ f t ▁ ˈθ i ▁ ˈw i t ▁ ˈf ɹ ɑ m ▁ ˈθ i ▁ ˈt ʃ æ f ▁ ˈɪ n ▁ ˈw æ t ▁ ˈθ e ɪ ▁ ˈh ɪ ɹ ▁ ˈæ n d ▁ ˈw æ t ▁ ˈθ e ɪ ▁ ˈɹ ɛ d ▁ .`
And your example above seems to have missed the punctuations. 

Hope this information helps, though I'm not sure if it is related.

**Comment by datquocnguyen** [comment]
As @godspirit00 mentioned, you should perform word segmentation on the raw text before running the preprocess.py script.
What computational resource you're having @cnlinxi ? What hyper-parameters did you use? The hyper-parameters might be sensitive for each dataset, you might do hyper-parameter tuning. 
You might also want to run the training process on a subset (1 hour) of your own dataset to see if there is something wrong.

**Comment by sourcesur** [comment]
@godspirit00 

> @datquocnguyen Sorry for the late reply. Unfortunately, no. I set the total number of epochs to 1000. It is now at Epoch 290 / 500k steps, and the output is still gibberish.
> 
>  1.mp4 
> The duration is also incorrect. GT is 8s but the gen is 2s.
> 
> Not sure if it is related, but the training encountered OOM at around epoch 250, probably when XPhoneBERT began to be finetuned. Then I edited the config and reduced the batch size, then resume the training. The number of steps jumped from 200k to 400k after resuming. If it's the reason, then I will have to restart the whole training right?

Hi, facing the same issue but my training had not been interrupted. 
I was training for 150 epochs on multispeaker dataset. XPoneBERT was frozen in the first 25% of epochs (37 epochs), at that time audio generation was fine but after the fine-tuning of XPhoneBERT started, the generated audio became gibberish (the same output as in your case), also generator and discriminator losses became nans.

Did your losses also become nans?
Did you just wait longer and the generation improved? How is the output now?

**Comment by lexkoro** [comment]
@sourcesur you might want to try lowering the learning rate only for xphonebert.

**Comment by godspirit00** [comment]
@sourcesur No, the losses were normal. 
Yes, after some more epochs, the generation is good. It is now at Epoch 587. The output now sounds good.

**Comment by sourcesur** [comment]
@thelinhbkhn2014 
Hi, I have started finetuning XPhoneBERT on the whole training process as suggested (no freezing at the start). I train with 1e-4 learning rate instead of 2e-4 since my batch size is 16 instead of 32. It's already Epoch 65/200 and I still get gibberish with wrong audio duration (e.g. 2s instead of 9s). Is this an expected behavior?

**Comment by thelinhbkhn2014** [comment]
I don't know why you can not train with your dataset like @godspirit00 . You can review your dataset, preprocessing data,.. and train longer.

**Comment by K2O7I** [comment]
Hi, have any of you successfully trained the model, I also had the same problem when training after 25% epochs, the model tends to get worse.

**Comment by wownaoh9** [comment]
> [@cnlinxi](https://github.com/cnlinxi) My training is at epoch 198/1000 now. I guess the finetuning of XPhoneBERT has not started so the output is still gibberish. I am waiting for it to go further.

How was the result after you retrained it?
